README.md
# DTree

--------------------------------------------------------------------------------
    DTree Is a Decision Tree which can do Classification and also Regression.
    Copyright (C) 2005 - Jordi Adell - TALP / UPC
-----------------------------------------------------------------------------------
                    GNU LESSER GENERAL PUBLIC LICENSE 
                        Version 3, 29 June 2007 

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
-------------------------------------------------------------------------------

DTree has been developed to generate decision trees which can
perform either classification and regression. It is supplied under the
GNU General Public License and is programed in C++.
	

#Download
========
  You can dowload the source code from GitHub here: https://github.com/jordi-adell/DTree


#Help
====
   It provides a binary for training trees and another one to test them,
The main code is inside a library (libDTree) that can be used by other
programs to train or use DTrees.

Code Documentation
     doc/ 

User Documentation:

     DTrain(v0.3 beta), training a decision tree

     DTrain [options]
     -d Data file
     -c Description file
     -t Output tree file
     -p Predictee fields (def: 0)
     -r Number of section for float fields (def: 10)
     -b Balance, minimum percentage of elements in a node (def: 0)
     -s Minimum number of elements in a leaf (stop criteria) (def: 50)
     -e Errortype. Options are: entropy, rmse, correlation, correct
     -v Print tree on screen
     -h Print this help

     ------------------------------------------------------------------------

     DTest(v0.3 beta) evaluating a decision tree

     DTest [options]
     -d Data file
     -c Description file
     -t Input tree file
     -o Output file
     -p Predictee fields (def: 0)
     -v Verbose predictions on stdout
     -h Print this help

     -------------------------------------------------------------------------

### Description File
   A text file containing a row for each field in the data file. Each row
must have two columns, on first one the name of the field is specified
and on second one the type of the field (float, enum).

Example:

  sepallength float
   
  sepalwidth float
   
  petallength float
  
  petalwidth float
  
  class enum


### Data file
   A text file with rows meaning elements and a column for each
field. Columns mush be separated by ' '.

* Example:
* 
   5.1 3.5 1.4 0.2 Iris-setosa

   4.9 3.0 1.4 0.2 Iris-setosa
   
   4.7 3.2 1.3 0.2 Iris-setosa
   
   4.6 3.1 1.5 0.2 Iris-setosa
   
   5.0 3.6 1.4 0.2 Iris-setosa
   
   5.4 3.9 1.7 0.4 Iris-setosa
   
   4.6 3.4 1.4 0.3 Iris-setosa
   
   5.0 3.4 1.5 0.2 Iris-setosa

### Output tree
   The name of the file where the tree will be written in.

### Predictee field
   Name of the files (as written in description file) that has to be
predictee by the tree.  Example: class

### Sections
   Float fields will be splitted in sections, in order to
   allow question such as Is petallength > 2.6. The bigger this number
   is the more questions the tree will handle, and the tree can better
   fit the training data. However it will make the training procedure
   slower and may lead to over-fitting.  

### Stopping Criteria (-s) 
   The number of elements in the leaf is used as stopping criteria. If no
more nodes with a number of elements bigger than this number can be
created the tree will stop growing.

### Error type 
   This option is used to choose the minimisation function. Default is
likelihood but also Root Mean Square Error and percentage of correctly
classified elements can be used.

If you need more help you can contact me at adellj@gmail.com


ChangeLog
06/03/2007 - Minor bug solved. Floiting point error  when dividing by zero in F-measure calculation.
08/04/2007 - Minor bug solved. Seg fault removed, variable clear in destructor produced a seg fault since varible didnt exist a that point.
08/04/2007 - Now data can be added element by element by means of the library.
08/04/2007 - Data matrix can be dump into a file.


src/DTree/DTrain.cpp
/*--------------------------------------------------------------------------------
    DTrain is based on DTree. It creates a Tree using DTree functions.
    Copyright (C) 2005 - Jordi Adell - TALP / UPC
-----------------------------------------------------------------------------------
                    GNU GENERAL PUBLIC LICENSE
		        Version 3, 29 June 2007

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
    MA  02110-1301, USA
-------------------------------------------------------------------------------*/

#include "DTree.h"

   
void disclaimer(const std::string &prog)
{
  cout << prog << " Copyright (C) 2005 - Jordi Adell" << endl;
  cout << "This program comes with ABSOLUTELY NO WARRANTY." << endl;
  cout << "This is free software, and you are welcome to redistribute it" << endl;
  cout << "under certain conditions." << endl;
}


void usage(string prog){
  disclaimer(prog);
  cout << endl;
  cout << "--------------------------------" << endl;
  cout << prog <<  "("<< DTREE_VERSION  << "), training a decision tree" << endl;
  cout << "--------------------------------" << endl;
  cout << prog << " [options]" << endl;
  cout << "\t-d Data file" << endl 
       << "\t-c Description file" << endl 
       << "\t-t Output tree file" << endl
       << "\t-p Predictee field name (def: 0)" << endl 
       << "\t-r Number of section for float fields (def: 10)" << endl 
       << "\t-b Balance, minimum percentage of elements in a node (def: 0.1)" << endl 
       << "\t-s Minumum number of elements in a leaf (stop criteria) (def: 50)" << endl 
       << "\t-e Errortype. Options are: entropy, rmse, correlation, correct" << endl
       << "\t-i Information file, with results" << endl
       << "\t-n N fold cross-validation (def:1)"  << endl
       << "\t-v Print tree on screen" << endl
       << "\t-h Print this help" << endl
       << endl;
}

// -- Main --------------------
int main(int argc, const char **argv){

  string datafile="", descfile="", treefile, option, spredictee, infofile;
  int predictee=0, minelements=50, section=10, ndata, folds=1;
  errType errortype=ENTROPY;
  float balance=0.0;
  bool printtree=false, leaveoneout=false;
  DTree tree;
  

  /*  cout   << "int       " << sizeof(int) << " " << numeric_limits<int>::max() << endl;
  cout   << "short int " << sizeof(short int) << " " << numeric_limits<short int>::max() << endl;
  cout   << "long int  " << sizeof(long int) << " " << numeric_limits<long int>::max() << endl;
  cout   << "float     " << sizeof(float) << " " << numeric_limits<float>::max() << endl;
  cout   << "double    " << sizeof(double) << " " << numeric_limits<double>::max() << endl;
  cout   << "long double " << sizeof(double) << " " << numeric_limits<double>::max() << endl;

  exit(0);
  */
  // Parsing the command line
  for (int i=1; i<argc; i++){
    if (argv[i][0] == '-')
      switch (argv[i][1]){
      case 'd':
	datafile=string(argv[++i]);
	break;
      case 'c':
	descfile=string(argv[++i]);
	break;
      case 'p':
	spredictee = string(argv[++i]);
	break;
      case 'b':
	balance = atof(argv[++i]);
	break;
      case 'r':
	section = atoi(argv[++i]);
	break;
      case 's':
	minelements = atoi(argv[++i]);
	break;
      case 't':
	treefile = string(argv[++i]);
	break;
      case 'i':
	infofile = string(argv[++i]);
	break;
      case 'v':
	printtree=true;
	break;
      case 'e':
	option = string(argv[++i]);
	if (option == "entropy")
	  errortype = ENTROPY;
	else if(option == "rmse")
	  errortype = RMSE;
	else if(option == "correct")
	  errortype = CORRECT;
	else if(option == "correlation")
	  errortype = CORRELATION;
	else
	  cerr << "WARNING!: Wrong type of optimization function. Entropy will be used  " << endl;
	
	break;
      case 'n':
	folds = atoi(argv[++i]);
	break;
      case 'l':
	leaveoneout = true;
	break;
      case 'h':
	usage(string(argv[0]));
	exit(0);
	break;
      default:
	cerr << "Wrong argument " << argv[i] << endl;
	exit (-1);
      }
    else{
      cerr << "Wrong argument " << argv[i] << endl;
      usage(string(argv[0]));
      exit(-1);
    }
  }

  if (argc < 6){
    cerr << "Wrong number of arguments" << endl;
    usage(string(argv[0]));
    exit(-1);
  }


  if (datafile.empty() || descfile.empty()){
    cerr << "Data and Description files must be provided" << endl;
    usage(string(argv[0]));
    exit (-1);
  }

  tree = DTree(predictee,balance,minelements,section,errortype);
  ndata = tree.dataLoad(datafile,descfile);
  if (leaveoneout) folds=ndata;

  vector<float> values;

  if (!spredictee.empty())
    tree.setPredictee(spredictee);

  cout << "------------------------" << endl
       << "DTree Training algorithm" << endl
       << "----------------------- " << endl;

  if (ndata <= 0){
    cerr << "ERROR: Loading data from " << datafile << " or  description file" << descfile << endl;
    exit(-1);
  }else
    cerr << ndata << " vectors loaded"  << endl;
  

  for (unsigned int i=0; i<folds; i++){
    if (folds <= 1){
      if(!tree.train()){
	cerr << "Problems on training " << endl;
	exit (-1);
      }
    }else{
      if(!tree.trainNfold((unsigned)folds,i)){
	cerr << "Problems on training " << endl;
	exit (-1);
      }
    }
    
    if (printtree)
      tree.dumpTree();
    
    if (!infofile.empty()){
      ofstream ofs(infofile.c_str(),ios::app);
      if (ofs.good())
	tree.dumpStatistics(ofs,true);      
      else
	cerr << "[ERROR]: Unable to open " << infofile << endl;
      ofs.close();
    }
    values.push_back(tree.dumpStatistics());      

    if (!treefile.empty())
      tree.saveTree(treefile);
    cout << "=======================" << endl;

  }
  if (folds > 1 && !values.empty()){
    float cumsum=0;
    
    size_t i=0;
    for (vector<float>::iterator it=values.begin(); it != values.end(); it++, i++){
      cumsum += *it;
      cout << i << "\t" << *it << endl;
    }
    cout << "MEAN:\t" << cumsum/values.size() << endl;
    cout << "------------------------" << endl;
  }


}


